lib/Sema/CSApply.cpp

```diff
@@ -2406,8 +2411,90 @@ namespace {
 
     // SWIFT_ENABLE_TENSORFLOW
     Expr *visitTFOp(ObjectLiteralExpr *expr) {
+      auto &ctx = cs.TC.Context;
+
+      auto diagnose = [&](SourceLoc loc, const Twine &message) {
+        cs.TC.diagnose(loc, diag::invalid_tfop, message.str());
+      };
+
       // All #tfop operands are RValues.
       expr->setArg(cs.coerceToRValue(expr->getArg()));
+
+      // We do most of the typechecking for #tfop here in CSApply instead of
+      // generating constraints for #tfop in CSGen because #tfop allows many
+      // different types of arguments and results that are inconvenient or
+      // impossible to generate constraints for. It's much easier to check #tfop
+      // after we know what all the argument and result types are.
+
+      // Check arguments.
+      auto *args = dyn_cast<TupleExpr>(expr->getArg());
+      if (!args) {
+        // CSGen has already ensured that the argument is a string in the
+        // single argument case.
+        return expr;
+      }
+
+      // CSGen has already ensured that the first argument is a string, so skip
+      // the first argument.
+      for (unsigned i = 1, e = args->getNumElements(); i != e; ++i) {
+        auto argExpr = args->getElement(i);
+        auto argType = cs.getType(argExpr);
+        auto argLoc = argExpr->getLoc();
+        auto argNameAndLowering = tf::GraphOperationInfo::decodeArgumentName(
+            args->getElementName(i).str());
+        if (!argNameAndLowering) {
+          diagnose(argLoc, "argument name has invalid modifier");
+          return nullptr;
+        }
+        auto argLowering = argNameAndLowering->second;
+        switch (argLowering) {
+        case tf::GraphOperationInfo::ArgumentLowering::Input:
+          // CSGen has already ensured that all inputs conform to
+          // TensorArrayProtocol.
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::NormalAttribute:
+          if (atc.classifyNormalAttribute(argType) ==
+              tf::AttributeTypeClassifier::Normal::Unsupported) {
+            diagnose(argLoc, "attribute requires Bool, Int64, Double, Float, "
+                             "array thereof, [TensorShape?], or Function, "
+                             "but got type '" + argType->getString() + "'");
+            return nullptr;
+          }
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::TensorAttribute:
```

Comment:
please confirm my understanding (this is also a design level issue that might be worth documenting in the PR description): for code simplicity, in eager mode (irgen) we do not support the `Const` node, so for uniformity of user code, a `#tfop("Const", ...)` will be rejected also in graph mode. But internally we can still synthesize these nodes (e.g. in deabstraction pass, we convert some Tensor initializers to Const nodes).

Are there any other TF ops that take tensor attrs, and could there be ones in the future? One theoretical reason could be that some people believe there could be a runtime performance advantage in taking attrs instead of inputs.

if this concern is valid, and there is a simple way to remove this restriction from the patch, please consider it, so that we can continue the discussion on tensor attrs out of the patch. I'm also ok with landing the patch as is, and continue the discussion.

Reply:
Yep, that understanding is right. I put it in the PR description.

Supporting tensor attrs in IRGen would require some pretty big refactoring. The issue is that graph lowering currently handles `$tensor` attrs by looking for a `$dtype` attr before the `$tensor` attr and a `$shape` attr immediately after the `$tensor` attr, which is possible in graph lowering mode because there is a single ordered list of const-evaluated attrs. In IRGen, there are two separate lists of attrs (dynamic & const-evaluated) with no ordering between them, so there is no way to look for dtype and shape attrs "before" and "after" the tensor attr. I think that the best way to address this would be to refactor the `$tensor` attr to take something like a tuple of `(dtype, tensor, shape)` so that it's all in one place. But that would be a pretty big change, so I will leave it out of this PR.

I'll file a ticket where we can track and discuss: https://bugs.swift.org/browse/SR-9249

Reply:
Could someone please explain why `Const` is not supported in eager mode?

Reply:
It's not supported in our IRGen because it has a $tensor attribute, which we don't support in IRGen because of the ordering thing I explained in my previous comment.

Reply:
Gotcha. You may want to clarify `$tensor` attribute is the reason why `Const` doesn't work, it wasn't obvious to me. (One could be misled into thinking "eagerness" is the reason why `Const` doesn't work.) 😃